bpf: make setting fifo policy non-fatal when probing hz

borkmann opened the pull request:
See commit msg.

pchaigno commented on bpf/cilium-probe-kernel-hz.c:
```diff
@@ -33,13 +33,26 @@ static int pin_to_cpu(int cpu)
+
+		if (ret_fifo || ret_nice)
+			printf("/* Note, could not set %s for probing KERNEL_HZ. */\n",
+			       ret_fifo && ret_nice ? "fifo and nice" : "fifo");
```
I tested this again with artificial load (processes with high priority on same CPU). When we can't set both fifo and nice, the program fails a lot more often, most of the time with a `probed jiffies mismatch` but also sometimes just with an incorrect HZ value (e.g., 300 or 1000).

I don't know what impact an incorrect HZ value would have on the datapath, but maybe we could also add a threshold on the warp value (when the HZ value is incorrect, warp is always 2-3-digits long)? Then again, I had to have several processes with highest priority running on the same core to see that happening, so probably quite unlikely.

borkmann replied:
Ok, to be on safe side, I've reworked everything. Will post in a bit.

pchaigno reviewed (approved):
@borkmann Could you also add the executable to `bpf/.gitignore`?

borkmann committed:
bpf: dump error on jiffie mismatch and add macro output

In case of an error in dump_kern_jiffies(), we need to propagate it so that
we exit with a non-success error code. Also move the macro output behind a
-m switch and add a non-macro output by default for cilium-probe-kernel-hz.

Also add cilium-probe-kernel-hz to gitignore.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium: add exec.Output() helper to only gather stdout

For cases where we don't care about a combined output, add a helper
to Cilium's exec library to only query stdout.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, bpf: improve and fix probing of kernel hz

This work removes probing of kernel hz from init.sh and and moves it into
the agent. I'll upstream a small bpftool change which reads CONFIG_HZ from
the underlying kernel config which Cilium agent can then pick up via probe
manager automatically:

  diff --git a/tools/bpf/bpftool/feature.c b/tools/bpf/bpftool/feature.c
  index 88718ee6a438..946e9b89467d 100644
  --- a/tools/bpf/bpftool/feature.c
  +++ b/tools/bpf/bpftool/feature.c
  @@ -378,6 +378,9 @@ static void probe_kernel_image_config(void)

                  /* test_bpf module for BPF tests */
                  "CONFIG_TEST_BPF",
  +
  +               /* Miscellaneous, e.g. for jiffies conversion */
  +               "CONFIG_HZ",
          };
          char *values[ARRAY_SIZE(options)] = { };
          struct utsname utsn;

If it is not present on the system, we then fall-back to run the probe
(cilium-probe-kernel-hz) and hold the result in the agent config. Either
way, we then emit the KERNEL_HZ when generating the node_config.h from
the agent. If kernel config or probing failed, we disable checking for
jiffie helper since there is no point due to missing KERNEL_HZ value.

Fixes: 8d5654fbfd31 ("bpf: probe and emit kernel hz to node_config")
Reported-by: Zang Li <zangli@google.com>
Reported-by: Benjamin Pineau <benjamin.pineau@datadoghq.com>
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

pchaigno reviewed (approved):
LGTM!
